docs/developer/architecture/kibana-platform-plugin-api.asciidoc

```diff
@@ -0,0 +1,765 @@
+
+|_stop_ |
+|===
+
+=== Integrating with other plugins
+
+Plugins can expose public interfaces for other plugins to consume. Like
+`core`, those interfaces are bound to the lifecycle functions `setup`
+and/or `start`.
+
+Anything returned from `setup` or `start` will act as the interface, and
+while not a technical requirement, all first-party Elastic plugins
+will expose types for that interface as well. 3rd party plugins
+wishing to allow other plugins to integrate with it are also highly
+encouraged to expose types for their plugin interfaces.
+
+*foobar plugin.ts:*
+
+[source,ts]
+----
+export type FoobarPluginSetup = ReturnType<Plugin['setup']>;
+export type FoobarPluginStart = ReturnType<Plugin['start']>;
+
+export class Plugin {
+  public setup() {
+    return {
+      getFoo() {
+        return 'foo';
+      },
+    };
+  }
+
+  public start() {
+    return {
+      getBar() {
+        return 'bar';
+      },
+    };
+  }
+}
+----
+
+Unlike core, capabilities exposed by plugins are _not_ automatically
+injected into all plugins. Instead, if a plugin wishes to use the public
+interface provided by another plugin, they must first declare that
+plugin as a dependency in their `kibana.json`.
+
+*demo kibana.json:*
+
+[source,json]
+----
+{
+  "id": "demo",
+  "requiredPlugins": ["foobar"],
+  "server": true,
+  "ui": true
+}
+----
+
+With that specified in the plugin manifest, the appropriate interfaces
+are then available via the second argument of `setup` and/or `start`:
+
+*demo plugin.ts:*
+
+[source,ts]
+----
+import { CoreSetup, CoreStart } from 'src/core/server';
+import { FoobarPluginSetup, FoobarPluginStop } from '../../foobar/server';
+
+interface DemoSetupPlugins { <1>
+  foobar: FoobarPluginSetup;
+}
+
+interface DemoStartPlugins {
+  foobar: FoobarPluginStart;
+}
+
+export class Plugin {
+  public setup(core: CoreSetup, plugins: DemoSetupPlugins) { <2>
+    const { foobar } = plugins;
+    foobar.getFoo(); // 'foo'
+    foobar.getBar(); // throws because getBar does not exist
+  }
+
+  public start(core: CoreStart, plugins: DemoStartPlugins) { <3>
+    const { foobar } = plugins;
+    foobar.getFoo(); // throws because getFoo does not exist
+    foobar.getBar(); // 'bar'
+  }
+
+  public stop() {},
+}
+----
+<1> The interface for plugin's dependencies must be manually composed. You can
+do this by importing the appopriate type from the plugin and constructing an
+interface where the property name is the plugin's ID.
+<2> These manually constructed typees should then be used to specify the type of
+the second argument to the plugin.
+<3> Notice that the type for setup and start are different. Plugin lifecycle
+functions can only access the APIs that are exposed during that lifecycle.
+
+=== Migrating legacy plugins to Kibana Platform
+
+In Kibana 7.10, support for legacy plugins was removed. See
+<<migrating-legacy-plugins>> for detailed information on how to convert existing
+legacy plugins to this new API.
+
+=== How to
```

Comment:
The sections here are kind of incoherent, it's just a dumping ground of information right now. I think this is probably ok for the first pass, but we should iterate on the organization here in the future.